### ISSUE TITLE
Crash on draw_screen

### ISSUE THREAD

**Issue by CameronLonsdale** [urwid/urwid#322]
##### Description:
In my project https://github.com/CameronLonsdale/MTP I have a list of edit boxes. In one edit box if I spam capital A I eventually get to the following traceback:

```
Traceback (most recent call last):
  File "/Users/clonsdale/Personal/manytime_interactive/venv/bin/mtp", line 11, in <module>
    load_entry_point('mtp', 'console_scripts', 'mtp')()
  File "/Users/clonsdale/Personal/manytime_interactive/cli.py", line 25, in main
    many_time_pad_attack(ciphertexts)
  File "/Users/clonsdale/Personal/manytime_interactive/manytime/__init__.py", line 19, in many_time_pad_attack
    interactive(ciphertexts, partial_key)
  File "/Users/clonsdale/Personal/manytime_interactive/manytime/interactive.py", line 191, in interactive
    Application(ciphertexts, Key(key)).run()
  File "/Users/clonsdale/Personal/manytime_interactive/manytime/interactive.py", line 143, in run
    self._loop.run()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 286, in run
    self._run()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 384, in _run
    self.event_loop.run()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 788, in run
    self._loop()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 816, in _loop
    self._entering_idle()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 777, in _entering_idle
    callback()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 572, in entering_idle
    self.draw_screen()
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/main_loop.py", line 587, in draw_screen
    self.screen.draw_screen(self.screen_size, canvas)
  File "/Users/clonsdale/Personal/manytime_interactive/venv/lib/python3.6/site-packages/urwid/raw_display.py", line 804, in draw_screen
    assert cs in [None, "0", "U"], repr(cs)
AssertionError: b'0'
```

##### Affected versions (if applicable)
- [ ] `master` branch (specify commit)
- [x] Latest stable version from `pypi`
- [ ] Other (specify source)

##### Steps to reproduce (if applicable)
Spam capital A on the first line after running `mtp examples/sample.ciphertexts`

##### Expected/actual outcome
This should not crash urwid, perhaps b'0' needs to be added to the list on line 804 in draw_screen ?

**Comment by tonycpsu** [comment]
I'm able to reproduce this, and the suggested fix does avoid this particular assert error, but I don't quite understand *why* a null would occur there.  @DoronCohen, was your experience with this bug also from using MTP, or something else?

**Comment by DoronCohen** [comment]
This was a long time ago. reproducing this will be very hard for me Im afraid

**Comment by CameronLonsdale** [comment]
From further testing it seems like the error comes from urwid trying to render something outside of the ascii range. In this example, it's character 0x9d, and it promptly goes from the first image, to the second. (This is with the modification in urwid to bypass the assert)

<img width="1009" alt="Screen Shot 2019-05-31 at 1 49 48 pm" src="https://user-images.githubusercontent.com/1342756/58680414-f9cab080-83aa-11e9-8e9b-105c9188cf77.png">

<img width="1129" alt="Screen Shot 2019-05-31 at 1 48 29 pm" src="https://user-images.githubusercontent.com/1342756/58680413-f9321a00-83aa-11e9-8340-bcace4d86387.png">

It just seems like urwid is protecting the terminal from getting funky renders. The best solution would be to modify my project to precheck for printable characters before getting urwid to render them.

**Comment by CameronLonsdale** [comment]
Although interestingly, when I type 'B' or other letters which a higher value, urwid handles this just fine by showing the '?' character instead of crashing. So it _should_ be hitting this case:

```
                if cs != 'U':
                    run = run.translate(UNPRINTABLE_TRANS_TABLE)
```
Cause then it would get translated into '?', but for some reason it's not.

**Comment by mfncooper** [comment]
There are two separate but related issues here.

First, let's look at the circumstances under which this happens. If a string to be rendered includes the SO (Shift Out, \x0e) character, which is a valid ASCII control character, but *not* the SI (Shift In, \x0f) character, this will cause Urwid to generate an invalid rle tuple, causing the crash. If a string includes *both* SO and SI characters, Urwid generates a valid rle tuple and there's no crash.

One issue here is that Urwid is interpreting strings that are passed to it. If a client's string happens to contain an SO character, Urwid will remove it, and that character will not be rendered at all. Urwid interprets the SO character as a switch to a DEC encoding, and SI as a switch back. There is no way to tell Urwid not to do this, so the workaround is to ensure that strings passed to Urwid for rendering never contain SO or SI control characters.

The actual cause of the crash is a bug in util.apply_target_encoding(), and this is where it becomes clear why the crash only happens when there is an SO without an SI.

If there are both SO and SI characters, we hit this line:

```python
            rle_append_modify(cout, (escape.DEC_TAG, len(sin)))
```

https://github.com/urwid/urwid/blob/master/urwid/util.py#L165

However, if there is an SO but not an SI, we hit this line:

```python
            rle_append_modify(cout, (escape.DEC_TAG.encode('ascii'), len(sin)))
```

https://github.com/urwid/urwid/blob/master/urwid/util.py#L159

Notice that, in the second case, the DEC_TAG value is being encoded. This is the bug; it causes the b'0' value seen in the rle.

Here's an illustration from the Python console:

```
>>> import urwid
>>>
>>> s1 = "hello \x0e happy \x0f world"
>>> urwid.apply_target_encoding(s1)
(b'hello  happy  world', [(None, 6), ('0', 7), (None, 6)])
>>>
>>> s2 = "hello \x0e sad world"
>>> urwid.apply_target_encoding(s2)
(b'hello  sad world', [(None, 6), (b'0', 10)])
>>>
```

Notice that the middle rle entry in the first item contains a '0', whereas the second rle entry in the second item contains a b'0' instead.

The fix is clear - simply remove the escape in the case in which SO appears without a matching SI.

Fixing this bug eliminates the crash, but it still leaves Urwid interpreting the SO and SI control characters, and removing them such that they are not rendered in any way. Thus it's still important for the client to ensure that any such characters are dealt with before passing strings to Urwid for rendering.

Incidentally, I do not believe that the changes in the PR attached to Issue #416 are necessary. The cs values are fine as characters ("U" or "0"), and there's no good reason to change everything to bytes.

**Comment by bugnano** [comment]
> Incidentally, I do not believe that the changes in the PR attached to Issue #416 are necessary. The cs values are fine as characters ("U" or "0"), and there's no good reason to change everything to bytes.

Did you try if the changes in #416 solve the problem? If they do, then I think that they *are* necessary, and the bug could have been fixed more than a year and a half ago by merging the PR.

The reason to change everything to bytes, I suspect that it's because of the Python2 to Python3 transition, where in Python2 everything was a string, but in Python3 we need to track which parts of the code deal with strings, and which ones deal with bytes. I did the change in #416 with that in mind: probably it wasn't clear after the Python3 transition that we're dealing with bytes here, so I simply made sure that the objects we're comparing are actually bytes and not strings.

**Comment by bugnano** [comment]
Given that we live in a Python3-only world for more than 2 years, I think that the enum solution is more than viable.
If #489 fixes the same problem as #416, then please, merge #489 and close #416, so that this long standing bug is fixed once and for all.
